abi/dynamic/reflect_marshal.go

```
@@ -0,0 +1,162 @@
+	if err := codec.LinearCodec.Unmarshal(data, dynamicValue); err != nil {
+		return "", fmt.Errorf("failed to unmarshal data: %w", err)
+	}
+
+	// Marshal the dynamic struct back to JSON
+	jsonData, err := json.Marshal(dynamicValue)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal struct to JSON: %w", err)
+	}
+
+	return string(jsonData), nil
+}
+
+func getReflectType(abiTypeName string, inputAbi abi.ABI, typeCache map[string]reflect.Type) reflect.Type {
+	switch abiTypeName {
+	case "string":
+		return reflect.TypeOf("")
+	case "uint8":
+		return reflect.TypeOf(uint8(0))
+	case "uint16":
+		return reflect.TypeOf(uint16(0))
+	case "uint32":
+		return reflect.TypeOf(uint32(0))
+	case "uint64":
+		return reflect.TypeOf(uint64(0))
+	case "int8":
+		return reflect.TypeOf(int8(0))
+	case "int16":
+		return reflect.TypeOf(int16(0))
+	case "int32":
+		return reflect.TypeOf(int32(0))
+	case "int64":
+		return reflect.TypeOf(int64(0))
+	case "Address":
+		return reflect.TypeOf(codec.Address{})
+	default:
+		if strings.HasPrefix(abiTypeName, "[]") {
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "[]"), inputAbi, typeCache)
+			return reflect.SliceOf(elemType)
+		} else if strings.HasPrefix(abiTypeName, "[") {
+			// Handle fixed-size arrays
+
+			sizeStr := strings.Split(abiTypeName, "]")[0]
+			sizeStr = strings.TrimPrefix(sizeStr, "[")
+
+			size, err := strconv.Atoi(sizeStr)
+			if err != nil {
+				return reflect.TypeOf((*interface{})(nil)).Elem()
+			}
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "["+sizeStr+"]"), inputAbi, typeCache)
+			return reflect.ArrayOf(size, elemType)
+		}
+		// For custom types, recursively construct the struct type
+
+		// Check if type already in cache
+		if cachedType, ok := typeCache[abiTypeName]; ok {
+			return cachedType
+		}
+
+		// Find the type in the ABI
+		abiType := findABIType(inputAbi, abiTypeName)
+		if abiType == nil {
+			// If not found, fallback to interface{}
```

Why? Please elaborate in the comment.

---

It makes much more sense to return errors here instead of just an empty value, I agree. Yeah, that was a dumb thing to do.

---

> that was a dumb thing to do

I don't actually think it was dumb as there could be reasons (e.g. you just want the JSON package to decide on the correct native type). I was just wondering what the reason was / prompting discussion about it.